### PR TITLE
Partial Ravager Rework - New Ravager Rage Ability

### PR DESCRIPTION
**Pull request by Surrealaser**
## About The Pull Request
1. Ravager gains the Rage ability:

**Rage
Duration:** 10 seconds
**Cooldown:** 2 Minutes
**Effect:** The Ravager bellows a demoralizing, earth shaking roar, regains plasma and gains extra speed, armor and slash damage in proportion to the % of its missing HP for the duration as well as slow, stagger and stun immunity. Plasma and Sunder are likewise restored. These benefits are further increased for each negative HP the Ravager has. The roar imposes 1-3 slow and stagger stacks (3 * Rage Power) on each non-allied living creature within range and earshot scaling in range with Rage strength (tile radius = 6 * Rage Power). 

**Details:**

Rage Power is calculated as 50% of the % difference between the Ravager's current and maximum hit points: (1 - (X / 350) ) * 0.5

So base Rage Power is 0.5 when the Ravager is at 0 HP, 0.64 when it's at -100 HP and 0.93 when it's at -300 HP.

Rage Power is further increased by 0.005 per point of negative HP, and is capped at 1.

Slash damage is increased by a % = to Rage Power.

Armor is increased by 40 * Rage Power

Speed is increased by 50% of Rage Power.

Plasma regained is a % = to Rage Power.

Sunder regained is equal to 50 * Rage Power (min 12.5, max 50)

2. Ignore Pain renamed to Endure, and its duration has been increased to 15 seconds up from 10, and cooldown increased from 40 seconds to 60 seconds, and plasma cost increased from 175 to 200. It now has an aura SFX and additional notifications. Ignore Pain grants immunity to the staggerslow effect from staggering munitions for its duration, and immunity to the stagger status in general. Instead of granting effective immortality, Endure lets the Ravager take up to 350 damage while at crit health before instantly dying. If Endure ends while the Ravager has normally lethal damage, its damage is reduced to barely non-lethal.

3. Added a xeno melee damage modifier var, allowing for global modification of slash damage for individual xenos.

4. Added statistics support for the Ravager abilities.

5. Charge cost decreased from 600 (lol) to 450. It still is impossible to combo Endure, Ravage and Charge all at once.

6. Ravage now also attacks structures and machinery in its area.

## Why It's Good For The Game

Adds an interesting high risk, high reward option to the Ravager that synergizes strongly with its existing kit, particularly Endure/Ignore Pain, and adds much needed depth to its kit.

Allows Ravage to target things other than humans only; it can now be used against hostile xenos, structures and machinery.

Fixes: https://github.com/tgstation/TerraGov-Marine-Corps/issues/5465

## Changelog
:cl:
add: Rage ability added.
balance: Ignore Pain renamed to Endure and its duration was increased from 10 to 20 seconds, cooldown increased from 40 seconds to 60 seconds, and plasma cost increased from 175 to 200; this also grants stagger immunity for its duration.
balance: Charge plasma cost decreased from 600 to 450.
refactor: Added statistics support to Ravager abilities. Added support for per xeno slash damage modification. Minor other refactors.
tweak: Ravage now targets all non-friendly living creatures as well as machinery and structures.
fix: Fixes Ignore Pain (now Endure) granting immortality. Instead the Ravager can take up to 350 damage while at crit health before instantly dying. If Endure ends while the Ravager has normally lethal damage, its damage is reduced to barely non-lethal.
/:cl:

### CODE REVIEW

**Comment by tgstation-server** [comment]
You currently have a negative Fix/Feature pull request delta of -3. Maintainers may close this PR at will. Fixing issues or improving the codebase will improve this score.

File: code/modules/mob/living/carbon/xenomorph/damage_procs.dm [hunk 1]
```diff
@@ -163,24 +163,30 @@
 
 
 /mob/living/carbon/xenomorph/adjustBruteLoss(amount, updating_health = FALSE)
+	if(stat == DEAD)  //Dead benos take no further damage.
+		return
+
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/damage_procs.dm:168]: remove this

File: code/modules/mob/living/carbon/xenomorph/damage_procs.dm [hunk 2]
```diff
@@ -163,24 +163,30 @@
...
 	var/list/amount_mod = list()
 	SEND_SIGNAL(src, COMSIG_XENOMORPH_BRUTE_DAMAGE, amount, amount_mod)
 	for(var/i in amount_mod)
 		amount -= i
 
-	bruteloss = clamp(bruteloss + amount, 0, maxHealth - xeno_caste.crit_health)
+	bruteloss = max(bruteloss + amount, 0)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/damage_procs.dm:175]: explain this

File: code/modules/mob/living/carbon/xenomorph/damage_procs.dm [hunk 3]
```diff
@@ -163,24 +163,30 @@
...
+	if(stat == DEAD) //Dead benos take no further damage.
+		return
+
 	var/list/amount_mod = list()
 	SEND_SIGNAL(src, COMSIG_XENOMORPH_BURN_DAMAGE, amount, amount_mod)
 	for(var/i in amount_mod)
 		amount -= i
 
-	fireloss = clamp(fireloss + amount, 0, maxHealth - xeno_caste.crit_health)
+	fireloss = max(fireloss + amount, 0)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/damage_procs.dm:189]: ditto

File: code/modules/projectiles/ammo_datums.dm [hunk 4]
```diff
@@ -88,11 +88,21 @@ GLOBAL_LIST_INIT(no_sticky_resin, typecacheof(list(/obj/item/clothing/mask/faceh
...
 			impact_message += "<span class='xenodanger'>Your fortified stance braces you against the impact.</span>"
 			return
+
+		if(xeno_victim.endure) //Endure allows us to ignore staggerstun.
+			impact_message += "<span class='xenodanger'>You endure the impact from [proj], shrugging off its effects.</span>"
+			return
+
+		if(xeno_victim.get_filter("ravager_rage_outline")) //Too angry to be affected; shut it down.
+			impact_message += "<span class='danger'>We're too angry to be staggered by the impact from [proj]!</span>"
+			return
```
Comment (TiviPlus) [inline code/modules/projectiles/ammo_datums.dm:98]: dont use a filter for this

File: code/modules/mob/living/carbon/xenomorph/castes/ravager/abilities_ravager.dm [hunk 5]
```diff
@@ -185,3 +245,160 @@
...
+			GW.add_filter("rage_outcry", 2, list("type" = "radial_blur", "size" = 0.07))
+			animate(GW.get_filter("rage_outcry"), size = 0.12, time = 5, loop = -1)
+			OT.add_filter("rage_outcry", 2, list("type" = "radial_blur", "size" = 0.07))
+			animate(OT.get_filter("rage_outcry"), size = 0.12, time = 5, loop = -1)
+			addtimer(CALLBACK(GW, /atom.proc/remove_filter, "rage_outcry"), 1 SECONDS)
+
+	for(var/turf/affected_tiles as() in RANGE_TURFS(rage_power_radius, X.loc))
+		affected_tiles.Shake(4, 4, 1 SECONDS) //SFX
+
+	for(var/mob/living/L in hearers(rage_power_radius, X)) //Roar that applies cool SFX and some soft CC
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/ravager/abilities_ravager.dm:332]: Dont spam hearers() its terribad for performance

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/__DEFINES/mobs.dm [hunk 6]
```diff
@@ -485,10 +485,16 @@ GLOBAL_LIST_INIT(xenoupgradetiers, list(XENO_UPGRADE_BASETYPE, XENO_UPGRADE_INVA
...
+#define RAVAGER_CHARGESPEED					2
+#define RAVAGER_CHARGESTRENGTH				2
+#define RAVAGER_CHARGEDISTANCE				4
+
+#define RAVAGER_ENDURE_DURATION				20 SECONDS
+#define RAVAGER_ENDURE_WARNING				0.66
+#define RAVAGER_ENDURE_HP_LIMIT				-350
+
+#define RAVAGER_RAGE_DURATION				10 SECONDS
+#define RAVAGER_RAGE_WARNING				0.7
```
Comment (SplinterGP) [inline code/__DEFINES/mobs.dm:497]: those should go to the caste datum, the non warning ones.

File: code/modules/mob/living/carbon/xenomorph/castes/ravager/abilities_ravager.dm (hunk 5 above)
Comment (ma44) [reply code/modules/mob/living/carbon/xenomorph/castes/ravager/abilities_ravager.dm:332]: You can cache the results of hearers() up in this file for the first time is used and reuse the results here.

It's still only 2 hearers() on a 2 minute cooldown so it can't be that expensive anyway.
